### PR TITLE
mixedversion: introduce preserve_downgrade_option_randomizer mutator

### LINKED ISSUE #111595: mixedversion: randomize point in which `preserve_downgrade_option` is reset
Currently, `mixedversion` tests always reset `cluster.preserve_downgrade_option` _after_ all nodes in the cluster are already running the new version. However, this is not the only way to upgrade: it's entirely valid to reset this cluster setting _before_ nodes are upgraded to use the new binary (or even while in mixed-version).

This randomization would have surfaced issues like #111539 in `mixedversion` tests. For discussion, see this [internal thread](https://cockroachlabs.slack.com/archives/C0KB9Q03D/p1696241428068959).

Jira issue: CRDB-31982

### PR DESCRIPTION
**Pull request by renatolabs**
This is the first `mutator` implementation added to the `mixedversion` framework: it randomizes the point in time when we reset the `cluster.preserve_downgrade_option` cluster setting in the test.

Fixes: #111595.

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
This PR introduces the first official mixedversion plan mutator: `preserve_downgrade_option_randomizer`. For motivation, see #111595. When this mutator is enabled, the point in which we call `RESET CLUSTER SETTING cluster.preserve_downgrade_option` is randomized.

I was able to reproduce the issue fixed by #111684 by running `acceptance/version-upgrade` in local mode enough times for the mutator to be enabled at the "right" time for the bug to manifest. If this was merged today, _some_ upgrade test would likely fail in the PR itself, or shortly after merge given the volume of test runs we get in CI and nightly tests.

Even though there will be some more follow-up work related to mutators, this PR concludes this series of PRs to introduce mutators. Thanks for reviewing!

### Context: mixedversion test mutators

This PR is part of a larger effort to introduce the concept of plan _mutators_ to the mixedversion framework. For initial discussion, see [internal thread](https://cockroachlabs.slack.com/archives/C01UDP33CCS/p1700252725103829).

In short, mutators are components that modify (i.e., mutate) the original test plan generated by the mixedversion planner. These changes are supposed to introduce new test scenarios in ways that do not interfere with the vast majority of tests. Tests will have the ability to individually enable or disable specific mutators as necessary; by default, all mutators will apply under some probability.

There is a working implementation of the entire idea in the [rc/mixedversion-plan-mutators](https://github.com/renatolabs/cockroach/tree/rc/mixedversion-plan-mutators) branch, in case it is helpful to see parts of the entire picture. That branch includes the proposed mutator API and an implementation of a mutator that, when merged, will close #111595.

#### History

* #115118: where we introduced the `upgradePlan` data structure and allows the upgrade to isolate the steps associated with each upgrade performed during the test.
* #117498: where we removed the `ID()` function from `singleStepProtocol`, allowing IDs to be assigned to steps later in 
the process, after the entire plan has been generated (and, now, mutations have been applied).
* #117993 is one of the pillars of this work, defining the interface to be implemented by mutators, as well as how the mutations they produce are applied to the original test plan.
* #118441 starts paving the way for the first mutators, with an API to disable them (in case a particular mutator is incompatible with a test), and by renaming the `finalize` step, foreshadowing the first (yet to be introduced) mutator implementation.
* #118666 makes another API change that brings user-provided mixed-version hooks and mixed-version functions implemented by the framework closer together. With those changes, the function to be provided by mutators looks and behaves just like a regular mixed-version hook in a test.

**Review by DarrylWong — COMMENTED** [review]
Played around with it in gceworker too for a bit and overall looks good to me, just two small questions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-:-NrH3dK9-PHzuwLbDilL:b-xrnhmf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-NrGiF3l3ZjkERwlVdKN:-NrGiF3l3ZjkERwlVdKO:bgt190j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b98c5abf9f13ce48ce23e9d35036bd3cecd341/pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go#L86)):*
> ```Go
> 					len(s.context.NodesInNextVersion()) == len(s.context.CockroachNodes)
> 			}) {
> 			step.context.Finalizing = true
> ```

I see that we are actually changing the `testPlan` here by modifying the `context`, but the actual mutations themselves need to be applied through an additional `applyMutations()` call. It's probably not an issue - I don't think we would want to generate mutations but not actually apply them, or call `Generate()` more than once - but just wanted to confirm.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer` line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-NrH3I0X0EZ7G4Kw615w:-NrH3I0X0EZ7G4Kw615x:b-q90nih) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b98c5abf9f13ce48ce23e9d35036bd3cecd341/pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer#L87)):*
> ```
> │   │   └── restart node 2 with binary version v23.2.4 (52) [stage=last-upgrade]
> │   ├── allow upgrade to happen by resetting `preserve_downgrade_option` (53) [stage=running-upgrade-migrations]
> │   ├── run "my mixed-version feature" (54) [stage=running-upgrade-migrations]
> ```

In this example, we are running user hooks while migrations are happening. Do we expect any current tests to break because of this? I know that tests can check `context.Finalizing()` but not sure how many test already do that.

**Review by renatolabs — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119225)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong and @srosenberg)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-NrGiF3l3ZjkERwlVdKN:-NrLYbvB9NiGC2Hf-kSP:bkka1tt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b98c5abf9f13ce48ce23e9d35036bd3cecd341/pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go#L86)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

I see that we are actually changing the `testPlan` here by modifying the `context`, but the actual mutations themselves need to be applied through an additional `applyMutations()` call. It's probably not an issue - I don't think we would want to generate mutations but not actually apply them, or call `Generate()` more than once - but just wanted to confirm.
</blockquote></details>

Good observation!

This is indeed a bit of an abstraction leak. The mutations API was designed with the purpose/assumption that mutator implementations won't change the mixed-version "context" in which the steps run. Changing the context means changing the binary versions running on each node at a given time, and whether the cluster is finalizing an upgrade.

I expect the vast majority of mutator implementations to not have/need to change that (including the planned uses we currently have for mutators -- fault injection, setting cluster settings, etc). However, this specific mutator is a bit special since it's moving the timing of the `allowUpgradeStep`. This timing _can_ have an impact on the mixed-version context, since it can change when the cluster is finalizing the upgrade.

TLDR: this does bypass the mutation API but it's a very contained change and similar approaches should not be needed in the future. We can think of a more uniform API in the future if this situation changes.

Long answer, but hopefully that makes sense.
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer` line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-NrH3I0X0EZ7G4Kw615w:-NrLZkxt65y6ZCSaZ-wo:b-ayyudx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b98c5abf9f13ce48ce23e9d35036bd3cecd341/pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer#L87)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

In this example, we are running user hooks while migrations are happening. Do we expect any current tests to break because of this? I know that tests can check `context.Finalizing()` but not sure how many test already do that.
</blockquote></details>

I don't expect this to be a problem. It's already the current behaviour that any function passed to `InMixedVersion` should expect to be called while the upgrade is finalizing, so this PR is not changing that. If, for some reason, they want to _not_ run in this situation, or behave differently, they can check the `Finalizing` field as you pointed out.

Note that this step you commented on is not affected -- the mutator did not apply in this specific upgrade. But it did apply in the last upgrade in this test plan. I made a slight tweak to the `debug` directive to also print the `Finalizing` field. As you can see, step 71 now knows that it's finalizing because we already reset `preserve_downgrade_option`, and all nodes have been upgraded to the latest binary.

**Review by DarrylWong — APPROVED** [review]
Alright, LGTM!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-:-NrMz5zU5El_2vk7e3DP:b9jd2wd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @srosenberg)

___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-NrGiF3l3ZjkERwlVdKN:-NrMxU922491BJQl_cvE:blas7zl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b98c5abf9f13ce48ce23e9d35036bd3cecd341/pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go#L86)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Good observation!

This is indeed a bit of an abstraction leak. The mutations API was designed with the purpose/assumption that mutator implementations won't change the mixed-version "context" in which the steps run. Changing the context means changing the binary versions running on each node at a given time, and whether the cluster is finalizing an upgrade.

I expect the vast majority of mutator implementations to not have/need to change that (including the planned uses we currently have for mutators -- fault injection, setting cluster settings, etc). However, this specific mutator is a bit special since it's moving the timing of the `allowUpgradeStep`. This timing _can_ have an impact on the mixed-version context, since it can change when the cluster is finalizing the upgrade.

TLDR: this does bypass the mutation API but it's a very contained change and similar approaches should not be needed in the future. We can think of a more uniform API in the future if this situation changes.

Long answer, but hopefully that makes sense.
</blockquote></details>

That makes perfect sense. Thanks for clarifying!
___
*[`pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer` line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119225#-NrH3I0X0EZ7G4Kw615w:-NrMy-8I4tL3kuyQZFMs:b1b7acn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3b98c5abf9f13ce48ce23e9d35036bd3cecd341/pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer#L87)):*
>should expect to be called while the upgrade is finalizing

Ahh yes you're completely right :smile: 

>tweak to the `debug` directive to also print the `Finalizing` field

Nice! In my first pass through I was having trouble reasoning when the Finalizing field was actually changing, this change helps a lot.

**Comment by renatolabs** [comment]
TFTR!

bors r=DarrylWong

File: pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+			}).
+			RandomStep(rng).
+			// Note that we don't attempt a concurrent insert because the
+			// selected step could be one that restarts a cockroach node,
+			// and `allowUpgradeStep` could fail in that situation.
+			InsertBefore(allowUpgradeStep{})
+
+		// Finally, we update the context associated with every step where
+		// all nodes are running the next verison to indicate they are in
+		// fact in `Finalizing` state. Previously, this would only be set
```
[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:77]: I'm not sure I understand the question, so I'll go for a longer answer in hopes it will cover what you are asking.

Before any test runs, we generate a "test plan" for the upgrade test. This includes the specific ordering of every event in the test: when nodes will restart, when user provided functions will be called, etc. The mixedversion context contains information relative to the mixed-version state the cluster is in, and is passed to user functions so that they can adapt based on current cluster state (e.g., pick a node running the older/newer version, do something different if migrations are running in the background, etc). All randomization happens during planning.

The mutators concept was added later, and is technically still part of the planning stage. Once we generate a base test plan, we non-deterministically apply a subset of the available mutators. When the mutator is chosen, we call `Generate` on the corresponding implementation. The mutations are then applied to the test plan, and the resulting test plan is passed to the runner, which blindly runs the steps in the test plan.

You probably already knew all of that, but I wanted to make sure we were on the same page. Specifically, I'm confused by the phrase

> `allowUpgradeStep` is guaranteed to run after the mutator

as `allowUpgradeStep` does _run_ but a mutator doesn't "run" in the same way -- it just changes the test plan.

`allowUpgradeStep` is a step in the test plan that resets `cluster.preserve_downgrade_option`. The time it runs is dependent on where the planner decides to place it in the test plan. By default (when the mutator introduced in this PR is not applied), we always reset once every node is restarted and is running the binary we want to upgrade to. When this mutator is enabled, we might reset it at different times (e.g., during a rollback, or halfway during our upgrade process).

When the mutator is _not_ enabled, we know that upgrade migrations will only run once `allowUpgradeStep` finishes. However, when the mutator _is_ enabled, migrations will run "as soon as all nodes are are on the next version". This is what the context update is capturing. If we didn't perform this update in the context, we would have steps that mistakenly thought they were _not_ finalizing when in fact they are. For a concrete example of this, see the thread above with Darryl and step 71 in the test added in this PR:

https://github.com/cockroachdb/cockroach/blob/2ae25505ce5edb4fbae9035abef9fe52f8ea5dfe/pkg/cmd/roachtest/roachtestutil/mixedversion/testdata/planner/preserve_downgrade_option_randomizer#L103-L108

Step 71 only knows it's "finalizing" because of this code that updates the context now that the `allowUpgradeStep` was moved.

*****

Hope some of this makes sense. If I misunderstood the question, happy to try to answer again.

[the remark this answers is not in the document]
Comment (srosenberg) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:77]: > When the mutator is not enabled, we know that upgrade migrations will only run once allowUpgradeStep finishes. However, when the mutator is enabled, migrations will run "as soon as all nodes are are on the next version"

Right. I think what I was missing is the fact that `allowUpgradeStep` doesn't set (or propagate) `Finalizing`. This happens in `finalizeUpgradeSteps`, which propagates `Finalizing=true` into the next steps. Since the mutator inserts a sequence of upgrade steps after `Plan`, `Finalizing=true` isn't propagated for this (new) sequence (since `finalizeUpgradeSteps` isn't invoked in this case). 

Note, if we had the knowledge all (transitive) dependents of `allowUpgradeStep`, then this context propagation could be moved directly into the step as opposed to being exposed to a mutator. More generally, we might want to think about what it would take to make _dependent_ sequences of steps, where the (dependence) ordering "closes" the whole sequence, i.e., invariants over the whole sequences are preserved.

In any case, thanks for a very detailed explanation!

[the remark this answers is not in the document]
Comment (renatolabs) [reply pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:77]: >  I think what I was missing is the fact that allowUpgradeStep doesn't set (or propagate) Finalizing.

Yeah, this would be tricky, as it wouldn't be safe to set `Finalizing` during `allowUpgradeStep` -- we might not be finalizing right after that step. `Finalizing` is a property of the mixed-version state, and `allowUpgradeStep` is only a part of that formula.

> then this context propagation could be moved directly into the step as opposed to being exposed to a mutator

IMO, a lot of the advantages of the current mutator API come from the fact that it _does_ have access to the context. For instance, in the mutator added in this PR, it's very easy to say "insert the allowUpgradeStep in any step during the rollback or last upgrade stages, and also choose a point in time when not all nodes are running the new binary". The context is what allows us to "query" the mixed-version state in this way.

But I also see your point and I think there would be advantages to steps managing the context. As I said above as well, I don't see us having to update the context in the vast majority of mutators so I'm not too concerned about this particular issue. We'll see if that changes.

**Comment by renatolabs** [comment]
No failures on master so far, I think we should be good to backport.

blathers backport 23.2 23.1
